cvat/apps/dataset_manager/bindings.py

```diff
@@ -2103,15 +2114,17 @@ def reduce_fn(acc, v):
                             attributes=attributes,
                         )
 
-                        tracks[track_id]['shapes'].append(track)
+                        if keyframe or outside:
+                            tracks[track_id]['shapes'].append(track)
+                            ending_tracks[track_id]['shapes'] = []
+                        else:
+                            ending_tracks[track_id]['shapes'].append(track)
```

Comment:
Probably, it should not be here, as we can't expect the parsed `item`s to be in the CVAT frame order. The order of track shapes is only established in the list in the line [2184](https://github.com/cvat-ai/cvat/blob/bb71350278816b563035b5b9f702ded7ac323177/cvat/apps/dataset_manager/bindings.py#L2184), so this code can possibly be added after that line.

Reply:
Instead of separating the shapes into `tracks` and `ending_tracks`, maybe I just dump everything into ending_tracks first, then sort and process it at line [2179](https://github.com/cvat-ai/cvat/blob/bb71350278816b563035b5b9f702ded7ac323177/cvat/apps/dataset_manager/bindings.py#L2179). Is this something you mean?

Reply:
Basically, yes, but I'm not sure why `ending_tracks` are needed in the first place. You can collect all the track shapes in `tracks`, then iterate over each track and build a new finalized list of shapes. In the loop we, basically, just need to infer the keyframes and add outside frames in the beginning of each track gap.

Please note that all `outside` frames must be keyframes as well. Each beginning of a non-outside shape group should be lead with a keyframe.

```

add outside = true here (where sequences of ":" start)
  v    v     v       v
[----------------------------]        current task or job frames
 |::|||::::||:::|||||::::::::         "|" is a shape, ":" - no shape

 ^  ^      ^    ^ 
add keyframe = true here (where sequences of "|" start)
```

Reply:
The reason I used `ending_tracks` is because I think it can help to skip some redundant reiteration of tracks. But I didn't consider the sequence of tracks at that time.